vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/DocumentOperationExecutor.java

```diff
@@ -0,0 +1,585 @@
+            default:
+                log.log(Level.WARNING, "Unexpected response outcome: " + outcome);
+            case ERROR:
+                return ERROR;
+        }
+    }
+
+    private void retry() {
+        maintain(throttled, delayed -> send(delayed.operation(), delayed.context()));
+    }
+
+    private void timeOut() {
+        maintain(timeouts, delayed -> delayed.context().error(TIMEOUT, "Timed out after " + defaultTimeout));
+    }
+
+    private void maintain(ConcurrentLinkedQueue<Delayed> queue, Consumer<Delayed> action) {
+        while ( ! Thread.currentThread().isInterrupted()) {
+            try {
+                Instant sleepUntil = null;
+                Iterator<Delayed> operations = queue.iterator();
+                while (operations.hasNext()) {
+                    Delayed operation = operations.next();
+                    // Already handled: remove and continue.
+                    if (operation.context().handled()) {
+                        operations.remove();
+                        continue;
+                    }
+                    // Ready for retry — remove from queue and send.
+                    if (operation.readyAt().isBefore(clock.instant())) {
+                        action.accept(operation);
+                        operations.remove();
+                        continue;
+                    }
+                    // Not yet ready for retry — note when to wake up again, unless already done.
+                    sleepUntil = sleepUntil != null ? sleepUntil : operation.readyAt();
+                }
+                long sleepMillis = sleepUntil != null ? sleepUntil.toEpochMilli() - clock.millis() : 10;
+                Thread.sleep(sleepMillis);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+
+    /** The executor will call <em>exactly one</em> callback <em>exactly once</em> for contexts submitted to it. */
+    private static class Context<T> {
+
+        private final AtomicBoolean handled = new AtomicBoolean();
+        private final BiConsumer<ErrorType, String> onError;
+        private final Consumer<T> onSuccess;
+
+        Context(BiConsumer<ErrorType, String> onError, Consumer<T> onSuccess) {
+            this.onError = onError;
+            this.onSuccess = onSuccess;
+        }
+
+        void error(ErrorType type, String message) {
+            if ( ! handled.getAndSet(true))
+                onError.accept(type, message);
+        }
+
+        void success(T result) {
+            if ( ! handled.getAndSet(true))
+                onSuccess.accept(result);
+        }
+
+        boolean handled() {
+            return handled.get();
+        }
+
+    }
+
+    /** Context for reacting to the progress of a visitor session. Completion signalled by an optional progress token. */
+    public static class VisitorContext extends Context<Optional<String>> {
```

Comment:
Consider renaming to `VisitOperationContext`?

Reply:
Sure!

Reply:
Well, it's not really an operation, but rather many of them.

Reply:
`VisitOperationsContext` then ;)

Reply:
Sold!